src/accountsdb/db.zig

```diff
@@ -2114,16 +2171,129 @@ pub const AccountsDB = struct {
+        };
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.accounts.deinit();
+        self.pubkeys.deinit();
+    }
+
+    pub fn reset(self: *Self) void {
+        self.accounts.clearRetainingCapacity();
+        self.pubkeys.clearRetainingCapacity();
+    }
+
+    pub fn cloneAndTrack(self: *Self, account_in_file: AccountInFile) Error!void {
+        // NOTE: this works because we mmap the account files - this will not work once we remove mmaps
+        const account = account_in_file.toAccount();
+
+        self.accounts.append(account) catch return Error.OutOfGeyserArrayMemory;
+        self.pubkeys.append(account_in_file.pubkey().*) catch return Error.OutOfGeyserArrayMemory;
+    }
+};
+
+pub const ValidateAccountFileError = error{
+    BinCountMismatch,
+    InvalidAccountFileLength,
+    OutOfReferenceMemory,
+} || AccountInFile.ValidateError || GeyserTmpStorage.Error;
+
+pub fn indexAndValidateAccountFile(
+    accounts_file: *AccountFile,
+    pubkey_bin_calculator: PubkeyBinCalculator,
+    bin_counts: []usize,
+    account_refs: *ArrayList(AccountRef),
+    geyser_storage: ?*GeyserTmpStorage,
+) ValidateAccountFileError!void {
+    var offset: usize = 0;
+    var number_of_accounts: usize = 0;
+
+    if (bin_counts.len != pubkey_bin_calculator.n_bins) {
+        return error.BinCountMismatch;
+    }
+
+    while (true) {
+        const account = accounts_file.readAccount(offset) catch break;
+        try account.validate();
+
+        if (geyser_storage) |storage| {
+            try storage.cloneAndTrack(account);
+        }
+
+        account_refs.append(.{
+            .pubkey = account.store_info.pubkey,
+            .slot = accounts_file.slot,
+            .location = .{
+                .File = .{
+                    .file_id = accounts_file.id,
+                    .offset = offset,
+                },
+            },
+        }) catch |err| switch (err) {
+            error.OutOfMemory => return error.OutOfReferenceMemory,
+        };
+
+        const pubkey = &account.store_info.pubkey;
+        const bin_index = pubkey_bin_calculator.binIndex(pubkey);
+        bin_counts[bin_index] += 1;
+
+        offset = offset + account.len;
+        number_of_accounts += 1;
+    }
+
+    const aligned_length = std.mem.alignForward(usize, accounts_file.length, @sizeOf(u64));
+    if (offset != aligned_length) {
+        return error.InvalidAccountFileLength;
+    }
+
+    accounts_file.number_of_accounts = number_of_accounts;
+}
+
+/// allocator which counts the number of times free is called. when count
+/// reaches 0, it will deinit the full arraylist. useful for when you want
+/// to allocate a large Arraylist and split it across multiple different
+/// ArrayLists -- alloc and resize are not implemented.
+///
+/// see `loadAndVerifyAccountsFiles` for an example of how to use this allocator
+const FreeCounterAllocator = struct {
     /// optional heap allocator to deinit the ptr on deinit
```

Comment:
Optional? 

Also on a side note, I have a slight preference to using 
`pub fn intoAllocator(self: *Self) std.mem.Allocator`  
or something similar over
`pub fn allocator(self: *Self) std.mem.Allocator` 
and thus avoiding the naming conflict. 

Given that struct fields have no accessibility modifiers I think that if we have a naming conflict between a struct field and a method it is likely that the method name could be more informative, the method must be doing 'something' otherwidse we would just access the field directly?

Reply:
Never mind, I see this is conforming to a pattern used by other allocators. Instead of `self_allocator` I think `backing_allocator` as used in RecycleFBA is better. On that note, should we move this allocator into `'utils/allocators.zig'`

Reply:
was originally thinking that too but the struct is so niche to accounts-db (eg, `references: ArrayList(AccountRef)`) i feel like its better here unless things change